### PR TITLE
Only discover neighbours for nodes known to not be end-nodes.

### PR DESCRIPTION
**Pull request by sesuncedu**
Both Mgmt_Lqi_req and Mgmt_Rtg_req must only be sent to a router or coordinator.

The ZigBeeNodeServiceDiscoverer did not verify this when discovering/updating neighbour information.

### CODE REVIEW

File: com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/app/discovery/ZigBeeNodeServiceDiscoverer.java [hunk 1]
```diff
@@ -461,6 +457,9 @@ private boolean requestActiveEndpoints() throws InterruptedException, ExecutionE
      * @throws InterruptedException
      */
     private boolean requestNeighborTable() throws InterruptedException, ExecutionException {
+        if (!supportsManagementLqi) {
```
Comment (cdjackson) [inline com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/app/discovery/ZigBeeNodeServiceDiscoverer.java]: Why is this necessary? If the task is removed due to the command not being supported, then it should not be necessary to check this again.

File: com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/app/discovery/ZigBeeNodeServiceDiscoverer.java [hunk 2]
```diff
@@ -515,6 +514,9 @@ private boolean requestNeighborTable() throws InterruptedException, ExecutionExc
      * @throws InterruptedException
      */
     private boolean requestRoutingTable() throws InterruptedException, ExecutionException {
+        if (!supportsManagementRouting) {
```
Comment (cdjackson) [inline com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/app/discovery/ZigBeeNodeServiceDiscoverer.java]: Why is this necessary? If the task is removed due to the command not being supported, then it should not be necessary to check this again.

File: com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/app/discovery/ZigBeeNodeServiceDiscoverer.java [hunk 3]
```diff
@@ -203,16 +203,12 @@ private void startDiscovery(Set<NodeDiscoveryTask> newTasks) {
         synchronized (discoveryTasks) {
             logger.debug("{}: Node SVC Discovery: starting new tasks {}", node.getIeeeAddress(), newTasks);
 
-            // Remove any tasks that we know are not supported by this device
-            if ((!supportsManagementLqi || node.getNodeDescriptor() != null
-                    && node.getNodeDescriptor().getLogicalType() == LogicalType.UNKNOWN)
-                    && newTasks.contains(NodeDiscoveryTask.NEIGHBORS)) {
+            // Remove router/coordinator-only tasks if the device is possibly an end node.
+            final boolean isPossibleEndDevice = isPossibleEndDevice();
+            if ((!supportsManagementLqi || isPossibleEndDevice)) {
```
Comment (cdjackson) [inline com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/app/discovery/ZigBeeNodeServiceDiscoverer.java]: By changing the logic around here it should not be needed to add the additional ```if supported``` checks in the tasks themselves.

File: com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/app/discovery/ZigBeeNodeServiceDiscoverer.java [hunk 4]
```diff
@@ -203,16 +203,12 @@ private void startDiscovery(Set<NodeDiscoveryTask> newTasks) {
...
                 newTasks.remove(NodeDiscoveryTask.NEIGHBORS);
             }
-            if ((!supportsManagementRouting || node.getNodeDescriptor() != null
-                    && (node.getNodeDescriptor().getLogicalType() == LogicalType.UNKNOWN
-                            || node.getNodeDescriptor().getLogicalType() == LogicalType.END_DEVICE))
-                    && newTasks.contains(NodeDiscoveryTask.ROUTES)) {
+            if ((!supportsManagementRouting || isPossibleEndDevice)) {
```
Comment (cdjackson) [inline com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/app/discovery/ZigBeeNodeServiceDiscoverer.java]: By changing the logic around here it should not be needed to add the additional ```if supported``` checks in the tasks themselves.

**Comment by cdjackson** [comment]
@sesuncedu I'd like to do a release of the framework and update to openHAB in the coming days. You've opened a number of PRs here - do you have time to address the comments? Some (like this one) should be pretty quick and it would really be good to get them closed out if you have time.

Thanks.

**Comment by sesuncedu** [comment]
I'll go check and clean on the ones I can cherry pick cleanly (I have these changes in the branch I am running, and I know that build cleanly, so there might be a commit with test changes that isn't in the pr branch.

**Comment by sesuncedu** [comment]
Oh, I remember this- it's a mock issue. I will see if I can pick the fix onto this branch.

**Comment by cdjackson** [comment]
Thanks. It would be really good if we can try and address PRs promptly and then we can avoid the hassle and wasted time of trying to cherry pick and resolve conflicts. I'll certainly try and review as quick as I can...

**Comment by sesuncedu** [comment]
Other  mock changes pushed. 
Have ten minutes to work on others before have to leave for footy.

File: com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/app/discovery/ZigBeeNodeServiceDiscoverer.java (hunk 1 above)
Comment (sesuncedu) [reply com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/app/discovery/ZigBeeNodeServiceDiscoverer.java]: Earlier safety check. I'll kill them

**Comment by cdjackson** [comment]
Thanks. I'll look at this shortly. Enjoy the game :)

File: com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/app/discovery/ZigBeeNodeServiceDiscoverer.java (hunk 3 above)
Comment (sesuncedu) [reply com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/app/discovery/ZigBeeNodeServiceDiscoverer.java]: Stet.

Since a possible end device may  actually not be an end device, whether or not the requests are supported may not be known a priori.  I think support is mandatory for ZR and ZC, but it's in the stack profile, so I'm not sure.

**Comment by cdjackson** [comment]
Thanks @sesuncedu - looks good.
